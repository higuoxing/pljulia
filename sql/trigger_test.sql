CREATE FUNCTION trigfunc_modcount() RETURNS trigger AS $$
if TD_event == "INSERT"
    TD_NEW["modcnt"] = 0
    return TD_NEW
elseif TD_event == "UPDATE"
    TD_NEW["modcnt"] = TD_OLD["modcnt"] + 1
    return TD_NEW
else
    return "OK"
end
$$ language pljulia;

CREATE TABLE mytab (num integer, description text, modcnt integer);
INSERT INTO mytab (num, description, modcnt) values (1, 'first', 0), (2, 'second', 0);

CREATE TRIGGER trig_mytab_modcount BEFORE INSERT OR UPDATE ON mytab
    FOR EACH ROW EXECUTE FUNCTION trigfunc_modcount('modcnt');

update mytab set description = 'first, modified once' where num = 1;
INSERT INTO mytab (num, description, modcnt) values (3, 'third', 3);

SELECT * FROM mytab;

DROP TRIGGER trig_mytab_modcount ON mytab;
DROP TABLE mytab;
DROP FUNCTION trigfunc_modcount;
