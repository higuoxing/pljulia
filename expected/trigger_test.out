CREATE FUNCTION trigfunc_modcount() RETURNS trigger AS $$
if TD_event == "INSERT"
    # args[1] == "modcnt"
    TD_NEW[args[1]] = 0
    return TD_NEW
elseif TD_event == "UPDATE"
    TD_NEW[args[1]] = TD_OLD[args[1]] + 1
    return TD_NEW
else
    return "OK"
end
$$ language pljulia;
CREATE TABLE mytab (num integer, description text, modcnt integer);
CREATE TRIGGER trig_mytab_modcount BEFORE INSERT OR UPDATE ON mytab
    FOR EACH ROW EXECUTE FUNCTION trigfunc_modcount('modcnt');
INSERT INTO mytab (num, description, modcnt) values (1, 'first', 0), (2, 'second', 0), (1, 'first', 1);
update mytab set description = 'first, modified once' where num = 1;
INSERT INTO mytab (num, description, modcnt) values (3, 'third', 3);
SELECT * FROM mytab;
 num |     description      | modcnt 
-----+----------------------+--------
   2 | second               |      0
   1 | first, modified once |      1
   1 | first, modified once |      1
   3 | third                |      0
(4 rows)

DROP TRIGGER trig_mytab_modcount ON mytab;
DROP TABLE mytab;
DROP FUNCTION trigfunc_modcount;
